Lot of -Wstrict-aliasing warnings

OS: `Gentoo`
GLideN64: f4b0755e2722d59fe1c20d7e470f7b8929dcae85
compiler: `gcc version 13.3.1 20240614 (Gentoo 13.3.1_p20240614 p17)`

When compiling GLideN64 there are a lot of `-Wstrict-aliasing` warnings, this can lead to unexpected behavior when building software and it would be great if it could be fixed.

It came up when updating RMG for the Gentoo repo. https://github.com/gentoo/gentoo/pull/38837#issuecomment-2395687366
```
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI4IA_RGBA4444(u16, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:55:39: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   55 |                 return IA88_RGBA4444(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF]);
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp:57:39: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   57 |                 return IA88_RGBA4444(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF]);
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI4IA_RGBA8888(u16, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:65:39: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   65 |                 return IA88_RGBA8888(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF]);
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp:67:39: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   67 |                 return IA88_RGBA8888(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF]);
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI4RGBA_RGBA5551(u16, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:75:43: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   75 |                 return RGBA5551_RGBA5551(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF]);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp:77:43: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   77 |                 return RGBA5551_RGBA5551(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF]);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI4RGBA_RGBA8888(u16, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:85:43: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   85 |                 return RGBA5551_RGBA8888(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF]);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp:87:43: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   87 |                 return RGBA5551_RGBA8888(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF]);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI8IA_RGBA4444(u16, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:123:31: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  123 |         return IA88_RGBA4444(*(u16*)&TMEM[(0x100 + color) & 0x1FF]);
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI8IA_RGBA8888(u16, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:129:31: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  129 |         return IA88_RGBA8888(*(u16*)&TMEM[(0x100 + color) & 0x1FF]);
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI8RGBA_RGBA5551(u16, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:135:35: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  135 |         return RGBA5551_RGBA5551(*(u16*)&TMEM[(0x100 + color) & 0x1FF]);
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI8RGBA_RGBA8888(u16, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:141:35: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  141 |         return RGBA5551_RGBA8888(*(u16*)&TMEM[(0x100 + color) & 0x1FF]);
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI16IA_RGBA8888(u16, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:196:27: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  196 |         const u16 col = (*(u16*)&TMEM[0x100 + (tex & 0xFF)]);
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI16IA_RGBA4444(u16, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:205:27: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  205 |         const u16 col = (*(u16*)&TMEM[0x100 + (tex & 0xFF)]);
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI4IA_RGBA4444_BG(u64*, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:306:39: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  306 |                 return IA88_RGBA4444(*(u16*)&TMEM[256 + (palette << 4) + (color4B & 0x0F)]);
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp:308:39: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  308 |                 return IA88_RGBA4444(*(u16*)&TMEM[256 + (palette << 4) + (color4B >> 4)]);
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI4IA_RGBA8888_BG(u64*, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:316:39: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  316 |                 return IA88_RGBA8888(*(u16*)&TMEM[256 + (palette << 4) + (color4B & 0x0F)]);
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp:318:39: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  318 |                 return IA88_RGBA8888(*(u16*)&TMEM[256 + (palette << 4) + (color4B >> 4)]);
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI4RGBA_RGBA5551_BG(u64*, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:326:43: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  326 |                 return RGBA5551_RGBA5551(*(u16*)&TMEM[256 + (palette << 4) + (color4B & 0x0F)]);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp:328:43: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  328 |                 return RGBA5551_RGBA5551(*(u16*)&TMEM[256 + (palette << 4) + (color4B >> 4)]);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI4RGBA_RGBA8888_BG(u64*, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:336:43: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  336 |                 return RGBA5551_RGBA8888(*(u16*)&TMEM[256 + (palette << 4) + (color4B & 0x0F)]);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp:338:43: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  338 |                 return RGBA5551_RGBA8888(*(u16*)&TMEM[256 + (palette << 4) + (color4B >> 4)]);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI8IA_RGBA4444_BG(u64*, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:371:31: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  371 |         return IA88_RGBA4444(*(u16*)&TMEM[256 + ((u8*)src)[x ^ (i << 1)]]);
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI8IA_RGBA8888_BG(u64*, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:376:31: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  376 |         return IA88_RGBA8888(*(u16*)&TMEM[256 + ((u8*)src)[x ^ (i << 1)]]);
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI8RGBA_RGBA5551_BG(u64*, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:381:35: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  381 |         return RGBA5551_RGBA5551(*(u16*)&TMEM[256 + ((u8*)src)[x ^ (i << 1)]]);
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI8RGBA_RGBA8888_BG(u64*, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:386:35: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  386 |         return RGBA5551_RGBA8888(*(u16*)&TMEM[256 + ((u8*)src)[x ^ (i << 1)]]);
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI16IA_RGBA8888_BG(u64*, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:432:27: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  432 |         const u16 col = (*(u16*)&TMEM[256 + (tex & 0xFF)]);
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/GLideN64/src/Textures.cpp: In function ‘u32 GetCI16IA_RGBA4444_BG(u64*, u16, u16, u8)’:
/tmp/GLideN64/src/Textures.cpp:441:27: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  441 |         const u16 col = (*(u16*)&TMEM[256 + (tex & 0xFF)]);
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[134/134] Linking CXX shared library plugin/Release/mupen64plus-video-GLideN64.so
```

--- a/src/Textures.cpp
+++ b/src/Textures.cpp
@@ -52,9 +52,9 @@ u32 GetCI4IA_RGBA4444(u16 offset, u16 x, u16 i, u8 palette)
 	const u8 color4B = Get4BitPaletteColor(offset, x, i);
 
 	if (x & 1)
-		return IA88_RGBA4444(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF]);
+		return IA88_RGBA4444(static_cast<u16>(TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF] & 0xFFFF));
 	else
-		return IA88_RGBA4444(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF]);
+		return IA88_RGBA4444(static_cast<u16>(TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF] & 0xFFFF));
 }
 
 u32 GetCI4IA_RGBA8888(u16 offset, u16 x, u16 i, u8 palette)
@@ -62,9 +62,9 @@ u32 GetCI4IA_RGBA8888(u16 offset, u16 x, u16 i, u8 palette)
 	const u8 color4B = Get4BitPaletteColor(offset, x, i);
 
 	if (x & 1)
-		return IA88_RGBA8888(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF]);
+		return IA88_RGBA8888(static_cast<u16>(TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF] & 0xFFFF));
 	else
-		return IA88_RGBA8888(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF]);
+		return IA88_RGBA8888(static_cast<u16>(TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF] & 0xFFFF));
 }
 
 u32 GetCI4RGBA_RGBA5551(u16 offset, u16 x, u16 i, u8 palette)
@@ -72,9 +72,9 @@ u32 GetCI4RGBA_RGBA5551(u16 offset, u16 x, u16 i, u8 palette)
 	const u8 color4B = Get4BitPaletteColor(offset, x, i);
 
 	if (x & 1)
-		return RGBA5551_RGBA5551(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF]);
+		return RGBA5551_RGBA5551(static_cast<u16>(TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF] & 0xFFFF));
 	else
-		return RGBA5551_RGBA5551(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF]);
+		return RGBA5551_RGBA5551(static_cast<u16>(TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF] & 0xFFFF));
 }
 
 u32 GetCI4RGBA_RGBA8888(u16 offset, u16 x, u16 i, u8 palette)
@@ -82,9 +82,9 @@ u32 GetCI4RGBA_RGBA8888(u16 offset, u16 x, u16 i, u8 palette)
 	const u8 color4B = Get4BitPaletteColor(offset, x, i);
 
 	if (x & 1)
-		return RGBA5551_RGBA8888(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF]);
+		return RGBA5551_RGBA8888(static_cast<u16>(TMEM[(0x100 + (palette << 4) + (color4B & 0x0F)) & 0x1FF] & 0xFFFF));
 	else
-		return RGBA5551_RGBA8888(*(u16*)&TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF]);
+		return RGBA5551_RGBA8888(static_cast<u16>(TMEM[(0x100 + (palette << 4) + (color4B >> 4)) & 0x1FF] & 0xFFFF));
 }
 
 u32 GetIA31_RGBA8888(u16 offset, u16 x, u16 i, u8 palette)
@@ -120,25 +120,25 @@ inline u8 Get8BitPaletteColor(u16 offset, u16 x, u16 i)
 u32 GetCI8IA_RGBA4444(u16 offset, u16 x, u16 i, u8 palette)
 {
 	const u8 color = Get8BitPaletteColor(offset, x, i);
-	return IA88_RGBA4444(*(u16*)&TMEM[(0x100 + color) & 0x1FF]);
+	return IA88_RGBA4444(static_cast<u16>(TMEM[(0x100 + color) & 0x1FF] & 0xFFFF));
 }
 
 u32 GetCI8IA_RGBA8888(u16 offset, u16 x, u16 i, u8 palette)
 {
 	const u8 color = Get8BitPaletteColor(offset, x, i);
-	return IA88_RGBA8888(*(u16*)&TMEM[(0x100 + color) & 0x1FF]);
+	return IA88_RGBA8888(static_cast<u16>(TMEM[(0x100 + color) & 0x1FF] & 0xFFFF));
 }
 
 u32 GetCI8RGBA_RGBA5551(u16 offset, u16 x, u16 i, u8 palette)
 {
 	const u8 color = Get8BitPaletteColor(offset, x, i);
-	return RGBA5551_RGBA5551(*(u16*)&TMEM[(0x100 + color) & 0x1FF]);
+	return RGBA5551_RGBA5551(static_cast<u16>(TMEM[(0x100 + color) & 0x1FF] & 0xFFFF));
 }
 
 u32 GetCI8RGBA_RGBA8888(u16 offset, u16 x, u16 i, u8 palette)
 {
 	const u8 color = Get8BitPaletteColor(offset, x, i);
-	return RGBA5551_RGBA8888(*(u16*)&TMEM[(0x100 + color) & 0x1FF]);
+	return RGBA5551_RGBA8888(static_cast<u16>(TMEM[(0x100 + color) & 0x1FF] & 0xFFFF));
 }
 
 u32 GetIA44_RGBA8888(u16 offset, u16 x, u16 i, u8 palette)
@@ -193,7 +193,7 @@ u32 GetI16_RGBA4444(u16 offset, u16 x, u16 i, u8 palette)
 u32 GetCI16IA_RGBA8888(u16 offset, u16 x, u16 i, u8 palette)
 {
 	const u16 tex = Get16BitColor(offset, x, i);
-	const u16 col = (*(u16*)&TMEM[0x100 + (tex & 0xFF)]);
+	const u16 col = (static_cast<u16>(TMEM[0x100 + (tex & 0xFF)] & 0xFFFF));
 	const u16 c = col >> 8;
 	const u16 a = col & 0xFF;
 	return (a << 24) | (c << 16) | (c << 8) | c;
@@ -202,7 +202,7 @@ u32 GetCI16IA_RGBA8888(u16 offset, u16 x, u16 i, u8 palette)
 u32 GetCI16IA_RGBA4444(u16 offset, u16 x, u16 i, u8 palette)
 {
 	const u16 tex = Get16BitColor(offset, x, i);
-	const u16 col = (*(u16*)&TMEM[0x100 + (tex & 0xFF)]);
+	const u16 col = (static_cast<u16>(TMEM[0x100 + (tex & 0xFF)] & 0xFFFF));
 	const u16 c = col >> 12;
 	const u16 a = col & 0x0F;
 	return (a << 12) | (c << 8) | (c << 4) | c;
@@ -303,9 +303,9 @@ u32 GetCI4IA_RGBA4444_BG(u64 *src, u16 x, u16 i, u8 palette)
 	u8 color4B = ((u8*)src)[(x >> 1) ^ (i << 1)];
 
 	if (x & 1)
-		return IA88_RGBA4444(*(u16*)&TMEM[256 + (palette << 4) + (color4B & 0x0F)]);
+		return IA88_RGBA4444(static_cast<u16>(TMEM[256 + (palette << 4) + (color4B & 0x0F)] & 0xFFFF));
 	else
-		return IA88_RGBA4444(*(u16*)&TMEM[256 + (palette << 4) + (color4B >> 4)]);
+		return IA88_RGBA4444(static_cast<u16>(TMEM[256 + (palette << 4) + (color4B >> 4)] & 0xFFFF));
 }
 
 u32 GetCI4IA_RGBA8888_BG(u64 *src, u16 x, u16 i, u8 palette)
@@ -313,9 +313,9 @@ u32 GetCI4IA_RGBA8888_BG(u64 *src, u16 x, u16 i, u8 palette)
 	u8 color4B = ((u8*)src)[(x >> 1) ^ (i << 1)];
 
 	if (x & 1)
-		return IA88_RGBA8888(*(u16*)&TMEM[256 + (palette << 4) + (color4B & 0x0F)]);
+		return IA88_RGBA8888(static_cast<u16>(TMEM[256 + (palette << 4) + (color4B & 0x0F)] & 0xFFFF));
 	else
-		return IA88_RGBA8888(*(u16*)&TMEM[256 + (palette << 4) + (color4B >> 4)]);
+		return IA88_RGBA8888(static_cast<u16>(TMEM[256 + (palette << 4) + (color4B >> 4)] & 0xFFFF));
 }
 
 u32 GetCI4RGBA_RGBA5551_BG(u64 *src, u16 x, u16 i, u8 palette)
@@ -323,9 +323,9 @@ u32 GetCI4RGBA_RGBA5551_BG(u64 *src, u16 x, u16 i, u8 palette)
 	u8 color4B = ((u8*)src)[(x >> 1) ^ (i << 1)];
 
 	if (x & 1)
-		return RGBA5551_RGBA5551(*(u16*)&TMEM[256 + (palette << 4) + (color4B & 0x0F)]);
+		return RGBA5551_RGBA5551(static_cast<u16>(TMEM[256 + (palette << 4) + (color4B & 0x0F)] & 0xFFFF));
 	else
-		return RGBA5551_RGBA5551(*(u16*)&TMEM[256 + (palette << 4) + (color4B >> 4)]);
+		return RGBA5551_RGBA5551(static_cast<u16>(TMEM[256 + (palette << 4) + (color4B >> 4)] & 0xFFFF));
 }
 
 u32 GetCI4RGBA_RGBA8888_BG(u64 *src, u16 x, u16 i, u8 palette)
@@ -333,9 +333,9 @@ u32 GetCI4RGBA_RGBA8888_BG(u64 *src, u16 x, u16 i, u8 palette)
 	u8 color4B = ((u8*)src)[(x >> 1) ^ (i << 1)];
 
 	if (x & 1)
-		return RGBA5551_RGBA8888(*(u16*)&TMEM[256 + (palette << 4) + (color4B & 0x0F)]);
+		return RGBA5551_RGBA8888(static_cast<u16>(TMEM[256 + (palette << 4) + (color4B & 0x0F)] & 0xFFFF));
 	else
-		return RGBA5551_RGBA8888(*(u16*)&TMEM[256 + (palette << 4) + (color4B >> 4)]);
+		return RGBA5551_RGBA8888(static_cast<u16>(TMEM[256 + (palette << 4) + (color4B >> 4)] & 0xFFFF));
 }
 
 u32 GetIA31_RGBA8888_BG(u64 *src, u16 x, u16 i, u8 palette)
@@ -368,22 +368,22 @@ u32 GetI4_RGBA4444_BG(u64 *src, u16 x, u16 i, u8 palette)
 
 u32 GetCI8IA_RGBA4444_BG(u64 *src, u16 x, u16 i, u8 palette)
 {
-	return IA88_RGBA4444(*(u16*)&TMEM[256 + ((u8*)src)[x ^ (i << 1)]]);
+	return IA88_RGBA4444(static_cast<u16>(TMEM[256 + ((u8*)src)[x ^ (i << 1)]] & 0xFFFF));
 }
 
 u32 GetCI8IA_RGBA8888_BG(u64 *src, u16 x, u16 i, u8 palette)
 {
-	return IA88_RGBA8888(*(u16*)&TMEM[256 + ((u8*)src)[x ^ (i << 1)]]);
+	return IA88_RGBA8888(static_cast<u16>(TMEM[256 + ((u8*)src)[x ^ (i << 1)]] & 0xFFFF));
 }
 
 u32 GetCI8RGBA_RGBA5551_BG(u64 *src, u16 x, u16 i, u8 palette)
 {
-	return RGBA5551_RGBA5551(*(u16*)&TMEM[256 + ((u8*)src)[x ^ (i << 1)]]);
+	return RGBA5551_RGBA5551(static_cast<u16>(TMEM[256 + ((u8*)src)[x ^ (i << 1)]] & 0xFFFF));
 }
 
 u32 GetCI8RGBA_RGBA8888_BG(u64 *src, u16 x, u16 i, u8 palette)
 {
-	return RGBA5551_RGBA8888(*(u16*)&TMEM[256 + ((u8*)src)[x ^ (i << 1)]]);
+	return RGBA5551_RGBA8888(static_cast<u16>(TMEM[256 + ((u8*)src)[x ^ (i << 1)]] & 0xFFFF));
 }
 
 u32 GetIA44_RGBA8888_BG(u64 *src, u16 x, u16 i, u8 palette)
@@ -429,7 +429,7 @@ u32 GetI16_RGBA4444_BG(u64 *src, u16 x, u16 i, u8 palette)
 u32 GetCI16IA_RGBA8888_BG(u64 *src, u16 x, u16 i, u8 palette)
 {
 	const u16 tex = ((u16*)src)[x^i];
-	const u16 col = (*(u16*)&TMEM[256 + (tex & 0xFF)]);
+	const u16 col = (static_cast<u16>(TMEM[256 + (tex & 0xFF)] & 0xFFFF));
 	const u16 c = col >> 8;
 	const u16 a = col & 0xFF;
 	return (a << 24) | (c << 16) | (c << 8) | c;
@@ -438,7 +438,7 @@ u32 GetCI16IA_RGBA8888_BG(u64 *src, u16 x, u16 i, u8 palette)
 u32 GetCI16IA_RGBA4444_BG(u64 *src, u16 x, u16 i, u8 palette)
 {
 	const u16 tex = ((u16*)src)[x^i];
-	const u16 col = (*(u16*)&TMEM[256 + (tex & 0xFF)]);
+	const u16 col = (static_cast<u16>(TMEM[256 + (tex & 0xFF)] & 0xFFFF));
 	const u16 c = col >> 12;
 	const u16 a = col & 0x0F;
 	return (a << 12) | (c << 8) | (c << 4) | c;

